src/client/client_api/blob_apis.rs

```
@@ -148,199 +157,117 @@ impl Client {
+            self.clone(),
+            (range.start..range.end + 1)
+                .clone()
+                .map(|i| all_keys[i].clone()),
+        )
+        .await;
+
+        if range.len() > encrypted_chunks.len() {
+            return Err(Error::Generic(format!(
+                "Missing chunks! Required {}, but we have {}.",
+                range.len(),
+                encrypted_chunks.len()
+            )));
         }
-    }
 
-    async fn delete_using_data_map(&self, data_map: DataMap) -> Result<(), Error> {
-        let blob_storage = BlobStorage::new(self.clone(), false);
-        let self_encryptor =
-            SelfEncryptor::new(blob_storage, data_map).map_err(Error::SelfEncryption)?;
+        self_encryption::decrypt_range(&secret_key, &encrypted_chunks, info.relative_pos, len)
+            .map_err(Error::SelfEncryption)
+    }
+
+    async fn get_chunks(
+        reader: Client,
+        keys: impl Iterator<Item = ChunkKey>,
+    ) -> Vec<EncryptedChunk> {
+        let tasks = keys.map(|key| {
+            let reader = reader.clone();
+            task::spawn(async move {
+                for i in 1..11 {
+                    let r_num: u8 = rand::random();
+                    let sleep_ms = i * r_num as u32;
+                    tokio::time::sleep(Duration::from_millis(sleep_ms as u64)).await;
+                    let result = reader.read_from_network(&key.dst_hash, false).await;
+                    if let Ok(chunk) = result {
+                        return Ok::<EncryptedChunk, Error>(EncryptedChunk {
+                            index: key.index,
+                            content: chunk.value().clone(),
+                        });
+                    }
+                }
+                Err(Error::NoResponse)
+            })
+        });
 
-        match self_encryptor.delete().await {
-            Ok(_) => Ok(()),
-            Err(error) => Err(Error::SelfEncryption(error)),
-        }
+        join_all(tasks)
```

this is trickier than the case of when storing a chunk
reading from network is not guaranteed to get a response, so it may return with an error or hang there infinitely.
join_all and swallow all error may cause many trouble to the later on debug or normal use.

---

We have timeouts further in there, so there will be no hanging here.

Swallowing the errors is just a way to compact them into one. There will be an error when we compare number of chunks required and the number we actually got.